

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Extract selected loci &mdash; odgi e809b89 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Sorting / Layouting" href="sorting_layouting.html" />
    <link rel="prev" title="Detect complex regions" href="detect_complex_regions.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> odgi
          

          
          </a>

          
            
            
              <div class="version">
                v0.5.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quick_start.html">Quick Start</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="exploratory_analysis.html">Exploratory analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="detect_complex_regions.html">Detect complex regions</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Extract selected <cite>loci</cite></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#synopsis">Synopsis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#steps">Steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#build-the-lipoprotein-a-graph">Build the Lipoprotein A graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#display-path-names">Display path names</a></li>
<li class="toctree-l4"><a class="reference internal" href="#obtain-the-lipoprotein-a-variants">Obtain the Lipoprotein A variants</a></li>
<li class="toctree-l4"><a class="reference internal" href="#visualize-the-sub-graph">Visualize the sub-graph</a></li>
<li class="toctree-l4"><a class="reference internal" href="#get-the-human-chr8-dataset">Get the Human chr8 dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="#extract-the-centromeres-beta-defensin-cluster">Extract the centromeres/beta-defensin cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="#do-something">DO SOMETHING</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="sorting_layouting.html">Sorting / Layouting</a></li>
<li class="toctree-l2"><a class="reference internal" href="translate_path_positions_between_graphs.html">Translate path positions between graphs</a></li>
<li class="toctree-l2"><a class="reference internal" href="remove_artifacts_and_complex_regions.html">Remove artifacts and complex regions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../commands.html">Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="../binding.html">Python Binding</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">odgi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorials.html">Tutorials</a> &raquo;</li>
        
      <li>Extract selected <cite>loci</cite></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/rst/tutorials/extract_selected_loci.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="extract-selected-loci">
<h1>Extract selected <cite>loci</cite><a class="headerlink" href="#extract-selected-loci" title="Permalink to this headline">¶</a></h1>
<div class="section" id="synopsis">
<h2>Synopsis<a class="headerlink" href="#synopsis" title="Permalink to this headline">¶</a></h2>
<p>Pangenome graphs model the full set of genomic elements in a gives species or clade. Nevertheless, downstream analyses
may require focusing on specific elements. <code class="docutils literal notranslate"><span class="pre">odgi</span></code> allows to extract <cite>loci</cite> of interest from the pangenome graph, for
easily working on them.</p>
</div>
<div class="section" id="steps">
<h2>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="build-the-lipoprotein-a-graph">
<h3>Build the Lipoprotein A graph<a class="headerlink" href="#build-the-lipoprotein-a-graph" title="Permalink to this headline">¶</a></h3>
<p>Assuming that your current working directory is the root of the <code class="docutils literal notranslate"><span class="pre">odgi</span></code> project, to construct an <code class="docutils literal notranslate"><span class="pre">odgi</span></code> file from the
<code class="docutils literal notranslate"><span class="pre">LPA</span></code> dataset in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi build -g test/LPA.gfa -o LPA.og
</pre></div>
</div>
<p>The command creates a file called <code class="docutils literal notranslate"><span class="pre">LPA.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains
13 contigs from 7 haploid human genome assemblies from 6 individuals plus the chm13 cell line. The contigs cover the
<a class="reference external" href="https://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000198670">Lipoprotein A (LPA)</a> locus, which encodes the
Apo(a) protein.</p>
</div>
<div class="section" id="display-path-names">
<h3>Display path names<a class="headerlink" href="#display-path-names" title="Permalink to this headline">¶</a></h3>
<p>The path’s names are:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi paths -i LPA.og -L
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chm13__LPA__tig00000001
HG002__LPA__tig00000001
HG002__LPA__tig00000005
HG00733__LPA__tig00000001
HG00733__LPA__tig00000008
HG01358__LPA__tig00000002
HG01358__LPA__tig00000010
HG02572__LPA__tig00000005
HG02572__LPA__tig00000001
NA19239__LPA__tig00000002
NA19239__LPA__tig00000006
NA19240__LPA__tig00000001
NA19240__LPA__tig00000012
</pre></div>
</div>
</div>
<div class="section" id="obtain-the-lipoprotein-a-variants">
<h3>Obtain the Lipoprotein A variants<a class="headerlink" href="#obtain-the-lipoprotein-a-variants" title="Permalink to this headline">¶</a></h3>
<p>To see variants for the two contigs of the <code class="docutils literal notranslate"><span class="pre">HG02572</span></code> sample, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>zgrep <span class="s1">&#39;^##&#39;</span> test/LPA.chm13__LPA__tig00000001.vcf.gz -v <span class="p">|</span> head -n <span class="m">5</span> <span class="p">|</span> cut -f <span class="m">1</span>-9,16,17 <span class="p">|</span> column -t
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#CHROM                   POS  ID      REF  ALT  QUAL  FILTER  INFO                        FORMAT  HG02572__LPA__tig00000001  HG02572__LPA__tig00000005
chm13__LPA__tig00000001  6    &gt;3&gt;6    T    A    60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      1                          0
chm13__LPA__tig00000001  134  &gt;6&gt;8    G    GT   60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      1                          0
chm13__LPA__tig00000001  156  &gt;8&gt;11   A    G    60    .       AC=1;AF=0.5;AN=2;LV=0;NS=2  GT      0                          1
chm13__LPA__tig00000001  252  &gt;11&gt;13  GT   G    60    .       AC=2;AF=1;AN=2;LV=0;NS=2    GT      1                          1
</pre></div>
</div>
<p>The pangenome graphs embed all the mutual relationship between the embedded genomes and their variation. In this example,
the variants are called respect to the <code class="docutils literal notranslate"><span class="pre">chm13__LPA__tig00000001</span></code> contig, which was used as reference path.</p>
<p>The insertion at position 136 (G &gt; GT) is present in only one of the  <code class="docutils literal notranslate"><span class="pre">HG02572</span></code>’s contig (<code class="docutils literal notranslate"><span class="pre">HG02572__LPA__tig00000001</span></code>).
To extract the sub-graph where this insertion falls, execute:</p>
<p>The instruction extracts:
- the node with ID 23 (<code class="docutils literal notranslate"><span class="pre">-n</span> <span class="pre">23</span></code>),
- the nodes reachable from this node following a single edge (<cite>-c 1</cite>) in the graph topology,
- the edges connecting all the extracted nodes, and
- the paths traversing all the extracted nodes.</p>
<p>To have basic information on the sub-graph, execute:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi stats -i LPA.21_23_G_GT.og -S
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>#length nodes   edges   paths
644     5       6       3
</pre></div>
</div>
<p>The extracted path’s names are:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>odgi paths -i LPA.21_23_G_GT.og -L
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>chm13__LPA__tig00000001:997-1640
HG02572__LPA__tig00000005:999-1641
HG02572__LPA__tig00000001:1035-1678
</pre></div>
</div>
<p>The sub-graph contains the contig used as reference in the <code class="docutils literal notranslate"><span class="pre">VCF</span></code> file, and the two <code class="docutils literal notranslate"><span class="pre">HG02572</span></code>’s contigs.</p>
</div>
<div class="section" id="visualize-the-sub-graph">
<h3>Visualize the sub-graph<a class="headerlink" href="#visualize-the-sub-graph" title="Permalink to this headline">¶</a></h3>
<p>To visualize the sub-graph, we can also use external tools as <a class="reference external" href="https://github.com/rrwick/Bandage">Bandage</a>, which
supports grpahs in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format. To covert the graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format in a graph in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, execute:</p>
<p>Then, open the <code class="docutils literal notranslate"><span class="pre">LPA.21_23_G_GT.gfa</span></code> file with <code class="docutils literal notranslate"><span class="pre">Bandage</span></code>.</p>
<img alt="../../_images/LPA.21_23_G_GT.png" src="../../_images/LPA.21_23_G_GT.png" />
<p>The image shows the graph topology, where each colored rectangle represents a node. In particular, three paths support
nodes with ID 21 and 23, and only one path supports the node with ID 22. The node with ID 22 represents in the graph the
additional nucleotide <code class="docutils literal notranslate"><span class="pre">T</span></code> presents in the <code class="docutils literal notranslate"><span class="pre">HG02572__LPA__tig00000001</span></code> contig as an insertion.</p>
</div>
<div class="section" id="get-the-human-chr8-dataset">
<h3>Get the Human chr8 dataset<a class="headerlink" href="#get-the-human-chr8-dataset" title="Permalink to this headline">¶</a></h3>
<p>Download the pangenome graph of the <a class="reference external" href="https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/scratch/2021_05_06_pggb/gfas/chr8.pan.gfa.gz">Human chromosome 8</a>
in <code class="docutils literal notranslate"><span class="pre">GFA</span></code> format, decompress it, and convert it to a graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -c https://s3-us-west-2.amazonaws.com/human-pangenomics/pangenomes/scratch/2021_05_06_pggb/gfas/chr8.pan.gfa.gz
gunzip chr8.pan.gfa.gz

odgi build -g chr8.pan.gfa -o chr8.pan.og
</pre></div>
</div>
<p>The last command creates a file called <code class="docutils literal notranslate"><span class="pre">chr8.pan.og</span></code>, which contains the input graph in <code class="docutils literal notranslate"><span class="pre">odgi</span></code> format. This graph contains
88 haploid, phased human genome assemblies from 44 individuals, plus the chm13 and GRCh38 reference genomes.</p>
</div>
<div class="section" id="extract-the-centromeres-beta-defensin-cluster">
<h3>Extract the centromeres/beta-defensin cluster<a class="headerlink" href="#extract-the-centromeres-beta-defensin-cluster" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="do-something">
<h3>DO SOMETHING<a class="headerlink" href="#do-something" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple" start="8">
<li><p>odgi viz/layout/bandage</p></li>
</ol>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="sorting_layouting.html" class="btn btn-neutral float-right" title="Sorting / Layouting" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="detect_complex_regions.html" class="btn btn-neutral float-left" title="Detect complex regions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Erik Garrison. Revision v0.5.1-e809b89.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>